### PR TITLE
[Inductor] Add backend hooks for C++ wrapper routing and device properties

### PR DESCRIPTION
**Pull request by srhust**
## Summary

Add backend extension points for C++ wrapper routing and device-property
metadata so out-of-tree accelerator backends can participate in these Inductor
paths without hardcoding their device names in core.

## Changes

- Add `DeviceOpOverrides.uses_gpu_cpp_wrapper()` with a default of `False`.
  CUDA and XPU override it to `True`, preserving the existing C++ wrapper
  routing behavior for in-tree backends.
- Use `uses_gpu_cpp_wrapper()` in both `GraphLowering` C++ wrapper decisions:
  the two-pass wrapper-codegen gate and the device selection used for JIT
  wrapper compilation.
- Treat unregistered devices conservatively as not using the GPU-style C++
  wrapper path.
- Add `DeviceInterface.get_multi_processor_count()`. The default implementation
  reads the standard `multi_processor_count` device-property field and asks
  backends with a different representation to override the method.
- Preserve the existing XPU behavior through `gpu_subslice_count` and the
  existing MTIA value of `64`.
- Make `DeviceProperties.create()` obtain the compute-unit count through
  `DeviceInterface.get_multi_processor_count()`.

## Motivation

Two Inductor paths currently encode backend-specific assumptions directly in
core: C++ wrapper routing recognizes CUDA/XPU by device name, and
`DeviceProperties.create()` contains device-specific handling for compute-unit
metadata.

For an out-of-tree backend, these differences are properties of the backend
rather than properties of Inductor itself. `DeviceOpOverrides` already owns
backend-specific codegen behavior, while `DeviceInterface` already owns
backend device properties.

Exposing the two decisions through those existing contracts lets registered
backends provide their behavior directly without adding out-of-tree device
names or property layouts to Inductor core.

## Test Plan

- CI: `python test/inductor/test_device_identity.py`
- Added device-independent tests for `get_multi_processor_count()`:
  standard-property behavior, missing-property error handling, the XPU
  `gpu_subslice_count` override, and the MTIA fixed value.
- Added a `DeviceProperties.create()` test using a fake `DeviceInterface` to
  verify that the compute-unit count is obtained through the interface
  contract.
- Added C++ wrapper-routing tests covering CUDA/XPU opt-in, CPU/MPS default
  behavior, an out-of-tree `DeviceOpOverrides` opt-in, the default
  out-of-tree behavior, and an unregistered device.
- The tests use fake interfaces / registered overrides and mocks where needed,
  so the new coverage does not require accelerator hardware.

## Notes

- `uses_gpu_cpp_wrapper()` is a backend codegen capability rather than a
  general device-classification predicate. Both relevant `GraphLowering`
  call sites use the same capability.
- The default behavior of both extension points is conservative:
  `DeviceOpOverrides` does not opt into GPU-style wrapper routing, while
  `DeviceInterface.get_multi_processor_count()` uses the standard device
  property when available.
- Existing CUDA/XPU C++ wrapper routing and XPU/MTIA compute-unit behavior are
  preserved.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by srhust** [comment]
> # This PR needs a `release notes:` label
> If your changes are user facing and intended to be a part of release notes, please use a label starting with `release notes:`.
> 
> If not, please add the `topic: not user facing` label.
> 
> To add a label, you can comment to pytorchbot, for example `@pytorchbot label "topic: not user facing"`
> 
> For more information, see https://github.com/pytorch/pytorch/wiki/PyTorch-AutoLabel-Bot#why-categorize-for-release-notes-and-how-does-it-work.

@pytorchbot label "topic: not user facing"

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes again

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Comment by srhust** [comment]
@jansel  Could you please rerun the failed CI job?
`linux-jammy-cuda13.2-cudnn9-py3.10-clang18 / build` failed during **Initialize containers** with `ErrImagePull / ImagePullBackOff`, before the build or tests started. This appears to be an infrastructure failure.

**Comment by jansel** [comment]
@claude review these changes again

**Comment by jansel** [comment]
@claude review these changes again

**Review by jansel — DISMISSED** [review]

**Comment by srhust** [comment]
@eellison  when you have a chance, could you please take a look at this PR?

**Review by JiasenTian — DISMISSED** [review]
LGTM

**Review by JiasenTian — DISMISSED** [review]

**Comment by srhust** [comment]
@atalman  The public ciflow tests on the current head (4a423a4) have not reproduced the internal failure, and the standard BC Lint/Lint/docs/pull workflows have also previously passed on this same head. Could you please share the failing internal test/job or the relevant error/stack trace that triggered the revert? I’d like to address the exact issue before relanding.

**Comment by srhust** [comment]
@jansel I've narrowed the scope of this PR substantially before requesting
another review.

#190326 already covers the general DeviceInterface-based GPU classification
work, including `is_gpu`, `exposes_streams`, `GPU_TYPES`, `get_gpu_type`, and
`device_need_guard`.

To avoid overlapping device-classification semantics, I've scoped this PR down
to the two remaining out-of-tree backend extension points:

1. `DeviceInterface.get_multi_processor_count()` for
   `DeviceProperties.create()`.
2. `DeviceOpOverrides.uses_gpu_cpp_wrapper()` for the two `GraphLowering`
   C++ wrapper routing decisions that historically special-case CUDA/XPU.

CUDA and XPU explicitly opt into the wrapper capability to preserve the
existing in-tree behavior, while an out-of-tree backend can opt in through its
registered `DeviceOpOverrides`. Both relevant `GraphLowering` call sites now
use the same capability.

Could you please take another look? Thanks.

**Comment by jansel** [comment]
@claude review this again; also give me a summary of the risk of relanding.  Do we have a good grasp of why this was reverted internally and the risks of it getting reverted again?

**Comment by jansel** [comment]
AI agenerated:


# Regression in PR #190615 — `has_gpu` gate in `codegen_with_cpp_wrapper`

## Summary

PR [#190615](https://github.com/pytorch/pytorch/pull/190615)
("[Inductor] Query DeviceInterface for device identity and capability")
was reverted because it broke Triton codegen for **non-CUDA/XPU accelerator
backends** (and, more subtly, for CPU when the CPU backend is Triton).

The culprit is a one-line change to the "does this graph have a GPU?" gate in
`GraphLowering.codegen_with_cpp_wrapper` (`torch/_inductor/graph.py`):

```diff
-        has_gpu = any(device in self.device_types for device in ["cuda", "xpu"])
+        has_gpu = any(ir.is_triton(device) for device in self.device_types)
```

That gate decides whether Inductor performs the **GPU two-pass AOTI /
cpp-wrapper compile** (materialize real inputs → autotune at compile time →
populate the Triton kernel/cubin cache):

```python
if has_gpu or needs_cpu_triton_two_pass:
    # extract real inputs, run compile-time autotuning, cache kernels ...
```

## Root cause

`has_gpu` is meant to mean *"this graph targets a real GPU that caches Triton
cubins"* (CUDA / XPU). The replacement predicate `ir.is_triton(device)` answers
a **different** question — *"is this device scheduled with `TritonScheduling`?"*:

```python
# torch/_inductor/ir.py
def is_triton(x) -> bool:
    device = get_device_type(x)
    if device in ["cpu", "cuda", "xpu"]:
        return getattr(config, f"{device}_backend") == "triton"
    device_scheduling = get_scheduling_for_device(device)
    if device_scheduling is None:
        return False
    from .codegen.triton import TritonScheduling
    return issubclass(device_scheduling, TritonScheduling)
```

So after the change `has_gpu` becomes `True` for cases it previously excluded:

1. **Any non-CUDA/XPU accelerator whose scheduler subclasses `TritonScheduling`.**
These backends were previously *excluded* from the GPU two-pass path and had
their own kernel-lowering / caching flow. Forcing them through the GPU
two-pass path changes the generated kernel metadata (including the stride
assumptions baked into the compiled kernels), which then no longer matches
the tensor strides in the graph — breaking downstream stride verification and
producing wrong/inconsistent codegen.
2. **Plain CPU graphs when `config.cpu_backend == "triton"`**, because
`is_triton("cpu")` returns `True` in that configuration. A CPU-only graph
would now be misclassified as `has_gpu`, entering the GPU cubin two-pass path
it was never meant to take (the narrow, intended CPU case is already handled
separately by `needs_cpu_triton_two_pass`).

In short: the change conflated **"is a GPU (cubin) device"** with **"uses the
Triton scheduler"**. Those are not the same set — CPU-with-Triton and
non-CUDA/XPU Triton accelerators live in the gap.

## Minimal repro (turn into a regression test)

The predicate difference is directly testable without any GPU:

```python
import torch
from torch._inductor import config, ir

# ir.is_triton() is True for CPU when the CPU backend is Triton...
with config.patch({"cpu_backend": "triton"}):
    assert ir.is_triton("cpu") is True

# ...but the OLD has_gpu gate excluded CPU:
device_types = {"cpu"}
old_has_gpu = any(d in ("cuda", "xpu") for d in device_types)
new_has_gpu = any(ir.is_triton(d) for d in device_types)

assert old_has_gpu is False
# Regression: the new predicate flips a CPU-only graph to "has_gpu"
assert new_has_gpu is True   # <-- this is the bug the PR introduced
```

The same flip happens for any registered non-CUDA/XPU device that uses
`TritonScheduling`: `old_has_gpu` is `False`, `new_has_gpu` is `True`, so the
device is incorrectly routed into the GPU two-pass compile.

An end-to-end regression test can AOT-compile a tiny model
(`torch._inductor.aot_compile`) for such a device with
`autotune_at_compile_time=False` and assert that the compile path / generated
kernel metadata is unchanged from the pre-#190615 behavior (e.g. that the GPU
two-pass branch is not taken for a CPU/non-cubin device).

## Suggested fix

Do **not** use `ir.is_triton` for this gate — it is about scheduler type, not
about being a cubin-caching GPU. Options, in order of preference:

1. **Leave the gate as an explicit GPU check** (smallest, behavior-preserving):
```python
has_gpu = any(device in ("cuda", "xpu") for device in self.device_types)
```
This line did not need to change for the device-generalization goal of the PR.
2. **If a hardcoded list must be avoided, add a dedicated capability bit** to
`DeviceInterface` that means exactly "this device uses the GPU Triton-cubin
two-pass compile" (e.g. `caches_triton_cubins()` / `needs_gpu_two_pass()`),
defaulting to `False`, and have only CUDA/XPU (and any GPU backend that truly
caches cubins) opt in:
```python
has_gpu = any(
    get_interface_for_device(device).needs_gpu_two_pass()
    for device in self.device_types
)
```
Crucially this must **exclude CPU and non-cubin accelerators** so their
existing kernel-lowering / caching flow is preserved.

The key point: the `has_gpu` gate must keep the *exact* set of devices it had
before (CUDA/XPU, i.e. real GPUs that cache Triton cubins). Any generalization
must be additive and opt-in, not derived from `ir.is_triton`, which silently
pulls in CPU-with-Triton and other-accelerator backends.

**Comment by jansel** [comment]
@claude based on the analysis above, update your opinion on reland safety

**Comment by jansel** [comment]
Suggestions from metamate on how to make this safer for MTIA:

Following up on the review's remaining blocking-ish item — making the MTIA/XPU
compute-unit change a **provable in-tree no-op** so the diff carries no behavioral
delta. Here's the concrete change I'd apply.

### Why `get_multi_processor_count()` needs the fallback

The pre-PR logic lived in `DeviceProperties.create()` and had a clear precedence:
**prefer the real field, fall back per-device.**

```python
try:
    multi_processor_count = props.multi_processor_count
except AttributeError:
    if device_type == "xpu":
        multi_processor_count = props.gpu_subslice_count
    elif device_type == "mtia":
        multi_processor_count = 64
    else:
        raise
```

The current override hardcodes `64` for MTIA unconditionally. That is only a no-op
if MTIA's device-properties object never exposes `multi_processor_count` — which
depends on the out-of-tree `getDeviceProperties()` implementation and isn't
observable from this repo. If that object *does* carry the field, the hardcode
silently drops the real value into reduction-split / max-autotune heuristics.

`multi_processor_count` feeds kernel-config heuristics, so a wrong value changes
generated MTIA kernels without any OSS CI signal — exactly the shape of an
internal-only regression.

### The fix (two lines, provably equivalent everywhere)

Preserve the original precedence with `getattr`-with-default so behavior is
identical regardless of what `getDeviceProperties()` returns:

```python
# MtiaInterface
@classmethod
def get_multi_processor_count(cls, device: torch.types.Device = None) -> int:
    # Prefer the real field if the properties object exposes it; otherwise
    # fall back to the historical constant. Matches the pre-existing
    # DeviceProperties.create() precedence exactly.
    return getattr(cls.get_device_properties(device), "multi_processor_count", 64)
```

```python
# XpuInterface
@classmethod
def get_multi_processor_count(cls, device: torch.types.Device = None) -> int:
    props = cls.get_device_properties(device)
    return getattr(props, "multi_processor_count", None) or props.gpu_subslice_count
```

This reproduces the old ladder in every case:

| `getDeviceProperties()` returns | old ladder | `getattr(..., 64)` | equivalent? |
| --- | --- | --- | --- |
| object **with** `multi_processor_count` | reads real value | reads real value | ✅ |
| object **without** the field | `AttributeError` → constant | constant | ✅ |
| a `dict` / stubbed properties | `AttributeError` → constant | constant | ✅ |

So it's a no-op for the in-tree stub *and* for any backend whose properties expose
the field — no dependency on out-of-tree behavior to reason about correctness.

### Tests

- `test_mtia_override` currently pins the constant. With the fallback it should
assert **both branches**: real field wins when present, constant when absent —
e.g. `mock.patch.object(MtiaInterface, "get_device_properties", return_value=SimpleNamespace())`
for the fallback case, and a `SimpleNamespace(multi_processor_count=...)` for the
real-field case.
- Add the in-tree no-op guarantee as a single CI-enforced invariant for the wrapper
gate, after `_initialize_device_op_overrides()`:
```python
assert {d for d in device_op_overrides_dict if _uses_gpu_cpp_wrapper(d)} == {"cuda", "xpu"}
```
- Keep the `cpu_backend="triton"` repro as a regression test so the old
`ir.is_triton` trap can't creep back into the gate.

### Naming (design nit, non-blocking)

`uses_gpu_cpp_wrapper()` reads as "is this a GPU wrapper device," but MPS
(`CppWrapperMps(CppWrapperGpu)`) and MTIA (`CppWrapperGpu`) both use a
`CppWrapperGpu` subclass yet must return `False`. A behavioral name like
`needs_two_pass_aoti_codegen()` (or a docstring that explicitly names MPS/MTIA as
`CppWrapperGpu` users that still return `False`) would stop the next backend author
from falling into the same trap that produced the earlier MPS AOTI error.

With the `getattr` fallbacks + the `{"cuda", "xpu"}` invariant assertion, the diff's
only new surface is two default-off, opt-in extension points, and every in-tree
device is a provable no-op.

**Comment by jansel** [comment]
@claude review again, update your reland safety analysis based on the above discussion and the new code

**Review by jansel — APPROVED** [review]
